### PR TITLE
Challenge 3 the tribonacci sequence

### PR DESCRIPTION
**Pull request by AnaiMurillo**
---
name: Challenge: The Tribonacci Sequence
about: A template PR for code review with a checklist
---

## Behavior
The "Tribonacci sequence" challenge is a twist on the famous Fibonacci sequence, where each number is the sum of the preceding three numbers.

`For example, 0, 1, 1, 2, 4, 7 …`

Write a function that returns the nth number in the Tribonacci sequence.

Example

```
tribonacci(5)
# Output: 4 (0, 1, 1, 2, 4)

tribonacci(7)
# Output: 13 (0, 1, 1, 2, 4, 7, 13)
```
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_tribonacci_sequence.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import unittest
```
Comment (likechrisss) [inline solutions/tests/test_tribonacci_sequence.py:15]: The test file does not contain a module-level docstring explaining:
1. _The purpose of the test module._
2. _Author information._

File: solutions/tests/test_tribonacci_sequence.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        self.assertEqual(tribonacci(6), 13)  # 0, 1, 1, 2, 4, 7, 13
+
+    def test_large_values(self):
+        """Test for larger values of n."""
+        self.assertEqual(tribonacci(10), 149)  # Verify correct calculation for T(10)
+        self.assertEqual(tribonacci(15), 3136)  # Verify correct calculation for T(15)
+
+    def test_edge_cases(self):
+        """Test edge cases."""
+        self.assertEqual(tribonacci(30), 29249425)  # Large n value
```
Comment (likechrisss) [inline solutions/tests/test_tribonacci_sequence.py:55]: Defensive assertions are missing.

File: solutions/the_tribonacci_sequence/main.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+def tribonacci(n: int) -> int:
+    """
+        Calculate the nth tribonacci number.
+
+        Args:
```
Comment (likechrisss) [inline solutions/the_tribonacci_sequence/main.py]: The docstring does not state that n must be:

1. A non-negative integer.
2. Raises an exception for invalid inputs.

**Review by likechrisss — CHANGES_REQUESTED** [review]
Great work but some things are missing for it to be perfect : 

1. You forgot the READMe file
2. The Function's Strategy Is Not Described in the Documentation (The current docstring briefly explains the Tribonacci sequence but doesn’t describe the iterative approach used in the function.)
3. It also lacks comments that explicitly describe the iterative approach used to compute the Tribonacci sequence.

File: solutions/the_tribonacci_sequence/main.py (hunk 3 above)
Comment (AnaiMurillo) [reply solutions/the_tribonacci_sequence/main.py]: I've updated the docstring to specify that `n` must be a non-negative integer and documented the `ValueError` exception for invalid inputs.

File: solutions/tests/test_tribonacci_sequence.py (hunk 2 above)
Comment (AnaiMurillo) [reply solutions/tests/test_tribonacci_sequence.py:55]: I added the defensive assertions in each test and also I made sure that this include as little logic as possible.

File: solutions/tests/test_tribonacci_sequence.py (hunk 1 above)
Comment (AnaiMurillo) [reply solutions/tests/test_tribonacci_sequence.py:15]: I have added the docstring to explain the two points you mentioned.

**Comment by AnaiMurillo** [comment]
@likechrisss
I have already made the requested changes, thanks for your recommendations!!! I will keep an eye out in case something is  missing.

**Review by likechrisss — APPROVED** [review]
Everything is now perfect.
Excellent job.
